### PR TITLE
Implement Unified Embeds for Forem-specific URLs

### PR DESCRIPTION
**Pull request by msarit**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This PR implements the [unified embed tag](https://github.com/forem/forem/issues/15099) for Forem-specific embeds.

## Related Tickets & Documents
closes https://github.com/forem/forem/issues/15565
closes https://github.com/forem/forem/issues/15569
closes https://github.com/forem/forem/issues/15580
closes https://github.com/forem/forem/issues/15583
Note: The issues for LinkTag and OrganizationTag were already closed by former PRs

## QA Instructions, Screenshots, Recordings
For links to test with, grab the following link types from your `localhost`:
- Link to User Profile (old method: `{% user <username> %}`)
- Link to Organization Profile (old method: `{% org <org-slug> %}`)
- Link to Listing (old method: `{% listing <listing-url> %}`)
- Link to Podcast Episode (old method: `{% podcast <episode-url> %}`)
- Link to Article (old method: `{% link <article-url-or-slug %}`)
- Link to Tag page (old method: `{% tag <tag-name> %}`)

As a user, create Forem-specific liquid tag using this format: `{% embed <forem-specific-url> %}`. The Liquid Tag should render properly.
Also create liquid tags using the old method (examples above). This should also work, as I am leaving the old implementation in place for now.

### UI accessibility concerns?
None

## Added/updated tests?
- [X] Yes

## [Forem core team only] How will this change be communicated?
- [X] I will share this change internally with the appropriate teams
Considering the project as a whole, I shall update the Liquid Tag documentation at the end. Since I am not changing the current implementation, the present documentation is still valid. Plus, I will avoid updating the docs piecemeal, and risk confusing users.

## [optional] Are there any post deployment tasks we need to perform?
None

### CODE REVIEW

File: spec/liquid_tags/unified_embed/registry_spec.rb [hunk 1]
```diff
@@ -3,12 +3,18 @@
 RSpec.describe UnifiedEmbed::Registry do
```
Comment (msarit) [inline spec/liquid_tags/unified_embed/registry_spec.rb:3]: Note: I alphabetized the specs in this file appropriately, so some unrelated specs may appear to have been written over, but they were not. 😅

File: app/liquid_tags/forem_tag.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  USER_ORG_REGEXP = %r{#{URL.url}/(?<name>[\w-]+)/?$}
+  POST_PODCAST_REGEXP = %r{#{URL.url}/(?<podcast>[\w-]+)/[\w-]+/?}
+  COMBINED_REGEXP = [USER_ORG_REGEXP, POST_PODCAST_REGEXP].freeze
+
+  def self.new(tag_name, input, parse_context)
+    link = process_input(input)
+    klass = determine_klass(link)
+    raise StandardError, "Invalid #{Settings::Community.community_name} URL." unless klass
+
+    klass.__send__(:new, tag_name, link, parse_context)
```
Comment (msarit) [inline app/liquid_tags/forem_tag.rb:12]: Note: When I used
```ruby
klass.new(tag_name, link, parse_context)
```
I got a `private method new` error. So I decided to use this `__send__` approach, which is also [used here](https://github.com/forem/forem/blob/0e900fd4dd07f31796251864b265eacb7cfc2364/app/liquid_tags/unified_embed/tag.rb#L37) in the `tag.rb` class.

Comment (jgaskins) [reply app/liquid_tags/forem_tag.rb:12]: Ooh, it makes a lot of sense to just route to it from inside `.new`! I love this!

File: app/liquid_tags/liquid_tag_base.rb [hunk 3]
```diff
@@ -51,6 +51,10 @@ def pattern_match_for(input, regex_options)
       .first
   end
 
+  def render_tag(klass, tag_name, link, parse_context)
+    klass.__send__(:new, tag_name, link, parse_context).render(nil)
+  end
+
```
Comment (jgaskins) [inline app/liquid_tags/liquid_tag_base.rb:57]: Looks like this is no longer needed with the `.new` change 🙂 

```suggestion
```

File: app/liquid_tags/forem_tag.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  def self.process_input(input)
+    stripped_input = ActionController::Base.helpers.strip_tags(input).strip
+    CGI.unescape_html(stripped_input)
+  end
+
+  def self.determine_klass(link)
+    return ListingTag if link.include?("#{URL.url}/listings/")
+    return TagTag if link.include?("#{URL.url}/t/")
```
Comment (jgaskins) [inline app/liquid_tags/forem_tag.rb:22]: Do these need to be `include?` or is `start_with?` still accurate?

```suggestion
    return ListingTag if link.start_with?("#{URL.url}/listings/")
    return TagTag if link.start_with?("#{URL.url}/t/")
```

**Review by jgaskins — APPROVED** [review]
Looks good! I also like the alphabetizing you did in the specs for the tag selection. :100:

The only thing that really stuck out to me was the `render_tag` method, but nothing in here is really blocking. I really like this idea for implementing a kind of generic "umbrella" embed for various more specific embeds.

File: app/liquid_tags/liquid_tag_base.rb (hunk 3 above)
Comment (msarit) [reply app/liquid_tags/liquid_tag_base.rb:57]: Did I not delete this? 🤦🏾‍♀️

File: app/liquid_tags/forem_tag.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
+module ForemTag
+  REGISTRY_REGEXP = %r{#{URL.url}/\w+(/[\w-]+)?(/[\w-]+)?}
+  USER_ORG_REGEXP = %r{#{URL.url}/(?<name>[\w-]+)/?$}
+  POST_PODCAST_REGEXP = %r{#{URL.url}/(?<podcast>[\w-]+)/[\w-]+/?}
+  COMBINED_REGEXP = [USER_ORG_REGEXP, POST_PODCAST_REGEXP].freeze
+
+  def self.new(tag_name, input, parse_context)
+    link = process_input(input)
+    klass = determine_klass(link)
```
Comment (jeremyf) [inline app/liquid_tags/forem_tag.rb:9]: This looks great!  I definitely want to see lots of tests of `determine_class` that's a lot of logic that is not quite verified.

**Review by jeremyf — CHANGES_REQUESTED** [review]
As I wrote earlier, I want to see that the logic is grabbing the correct "forem specific" tag class.

File: app/liquid_tags/forem_tag.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
+module ForemTag
+  REGISTRY_REGEXP = %r{#{URL.url}/\w+(/[\w-]+)?(/[\w-]+)?}
```
Comment (jeremyf) [inline app/liquid_tags/forem_tag.rb:2]: I suspect that we really need only check if text matches the URL.url.  All other "slugs" will get picked up by the `determine_klass` method.

**Review by jeremyf — APPROVED** [review]

**Comment by jeremyf** [comment]
Great stuff!

**Pull request merged**